File: backend/app/schemas/ai.py
"""DeepSeek AI agent schemas."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import Field

from app.schemas.common import SchemaBase


class AiSettingsResponse(SchemaBase):
    configured: bool
    model: str
    base_url: str
    api_key_masked: str | None = None
    agent_name: str = "SNR Dev"
    updated_at: str | None = None


class AiSettingsUpdateRequest(SchemaBase):
    api_key: str | None = None
    model: str | None = None
    agent_name: str | None = None
    clear: bool = False


class AiChatMessage(SchemaBase):
    role: Literal["user", "assistant", "system", "tool"]
    content: str
    name: str | None = None
    tool_call_id: str | None = None


class AiPendingAction(SchemaBase):
    id: str
    type: Literal["write_file", "terminal", "mkdir", "write_files", "create_database", "drop_database", "run_sql", "run_mongo"]
    reason: str
    path: str | None = None
    content: str | None = None
    command: str | None = None
    cwd: str | None = None
    app_id: str | None = None
    root_id: str | None = None
    token: str
    preview: str | None = None
    critical: bool = False
    files: list[dict[str, Any]] | None = None
    database: dict[str, Any] | None = None
    edits: list[dict[str, Any]] | None = None
    patch: bool = False


class AiToolTrace(SchemaBase):
    name: str
    arguments: dict[str, Any] = Field(default_factory=dict)
    result_preview: str


class AiChatRequest(SchemaBase):
    message: str = Field(min_length=1, max_length=32000)
    history: list[AiChatMessage] = Field(default_factory=list, max_length=120)
    surface: Literal["files", "terminal", "editor", "dashboard", "studio", "portal"] = "files"
    path: str | None = None
    app_id: str | None = None
    root_id: str | None = None
    cwd: str | None = None
    session_id: str | None = None
    file_content: str | None = Field(default=None, max_length=400_000)
    original_content: str | None = Field(default=None, max_length=400_000)


class AiUsageStats(SchemaBase):
    prompt_tokens: int = 0
    completion_tokens: int = 0
    total_tokens: int = 0
    weighted_tokens: int = 0
    credits_charged: int | None = None
    credits_remaining: int | None = None
    tokens_remaining: int | None = None
    tokens_per_credit: int | None = None


class AiChatResponse(SchemaBase):
    reply: str
    pending_actions: list[AiPendingAction] = Field(default_factory=list)
    tool_traces: list[AiToolTrace] = Field(default_factory=list)
    configured: bool = True
    session_id: str | None = None
    usage: AiUsageStats | None = None


class AiApplyActionRequest(SchemaBase):
    token: str = Field(min_length=10)
    confirm_password: str | None = Field(default=None, max_length=128)


class AiSessionSummary(SchemaBase):
    id: str
    title: str
    surface: str
    path: str | None = None
    app_id: str | None = None
    root_id: str | None = None
    message_count: int = 0
    created_at: str | None = None
    updated_at: str | None = None


class AiSessionDetail(SchemaBase):
    id: str
    title: str
    surface: str
    path: str | None = None
    app_id: str | None = None
    root_id: str | None = None
    messages: list[dict[str, Any]] = Field(default_factory=list)
    created_at: str | None = None
    updated_at: str | None = None


class AiSessionCreateRequest(SchemaBase):
    surface: Literal["files", "terminal", "editor", "dashboard", "studio", "portal"] = "files"
    title: str | None = None
    path: str | None = None
    app_id: str | None = None
    root_id: str | None = None


class AiSessionListResponse(SchemaBase):
    sessions: list[AiSessionSummary] = Field(default_factory=list)

File: frontend/src/types/ai.ts
export interface AiSettings {
  configured: boolean
  model: string
  base_url: string
  api_key_masked: string | null
  agent_name?: string
  updated_at: string | null
}

export interface AiChatMessage {
  role: 'user' | 'assistant' | 'system' | 'tool'
  content: string
  name?: string | null
  tool_call_id?: string | null
}

export interface AiPendingAction {
  id: string
  type: 'write_file' | 'terminal' | 'mkdir' | 'write_files' | 'create_database' | 'drop_database' | 'run_sql' | 'run_mongo'
  reason: string
  path?: string | null
  content?: string | null
  command?: string | null
  cwd?: string | null
  app_id?: string | null
  root_id?: string | null
  token: string
  preview?: string | null
  critical?: boolean
  files?: Array<{ path?: string; absolute_path?: string; content?: string }> | null
  database?: Record<string, unknown> | null
  edits?: Array<{ old_text: string; new_text: string }> | null
  patch?: boolean
}

export interface AiToolTrace {
  name: string
  arguments: Record<string, unknown>
  result_preview: string
}

export interface AiChatResponse {
  reply: string
  pending_actions: AiPendingAction[]
  tool_traces: AiToolTrace[]
  configured: boolean
  session_id?: string | null
}

export interface AiPanelMessage {
  id: string
  role: 'user' | 'assistant'
  content: string
  pending?: AiPendingAction[]
  traces?: AiToolTrace[]
}

export interface AiSessionSummary {
  id: string
  title: string
  surface: string
  path?: string | null
  app_id?: string | null
  root_id?: string | null
  message_count: number
  created_at?: string | null
  updated_at?: string | null
}

export interface AiSessionDetail {
  id: string
  title: string
  surface: string
  path?: string | null
  app_id?: string | null
  root_id?: string | null
  messages: Array<{ id?: string; role: string; content: string }>
  created_at?: string | null
  updated_at?: string | null
}
